### PR TITLE
feat(useCellRangeSelection): Ability to select single cells and Multi-range cells

### PR DESCRIPTION
**Pull request by 07harish**
Ability to select single cells (ctrl + click) and Multi-range cells (click and drag)

**State**:

`currentSelectedCellIds (map)` : Cell Ids of a current selected range
`selectedCellIds (map)` : Contains all selected cell Ids
`isSelectingCells (boolean)` : If the range is currently being selected

`startCellSelection and endCellSelection`: During range selection, these are the start cell Id and end cell Id, This is not used to setState cells that come between `startCellSelection and endCellSelection`. This might be of less requirement for the user. But I have used it for internal implementation

**Instance**:

`cellIdSplitBy (string)` : Cell Ids are split by using this string or  by default it is `_col_row_`
`cellsById (map)` : Same as `rowsById `
`getCellsBetweenId (Fn)` :  `(startCellId, endCellId) => map`  Returns all cell Ids between Range ( between startcell and endcell Ids)
`setSelectedCellIds(Fn)`: disptaches the action to set selectedCellIds

**Usage (Ux**): 
- Just `drag` is required to select first range
- `ctrl + click` to select single(and multiple) cells and clicking again will deselect the single cell
- `ctrl + drag` to select multiple-range cells

> setState for Range Selection can be done using 
> `instance.getCellsBetweenId(startCellId, endCellId)` which returns all cell Ids in a range and then `instance.setSelectedCellIds`
> 

Thanks, @gargroh 😃  for giving me a direction to work on this. Let me know your feedback and if any changes are needed.

### CODE REVIEW

**Comment by 07harish** [comment]
I will write tests soon for this.

File: examples/cell-range-selection/src/App.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                      {cell.render('Cell')}
+                    </td>
+                  )
+                })}
+              </tr>
+            )
+          })}
+        </tbody>
+      </table>
+    </>
```
Comment (gargroh) [inline examples/cell-range-selection/src/App.js:119]: Please add `selectedCellIds` state for display below table

File: examples/cell-range-selection/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app) and Rescripts.
+
+
+ <!-- Add codesanbox link once deployed -->
```
Comment (gargroh) [inline examples/cell-range-selection/README.md:4]: Before merge pls update this, mentioning so that it is not missed

**Comment by gargroh** [comment]
@07harish , mentioning some points let me know your thoughts on it - 

* Can we have documentation of plugin something like https://github.com/gargroh/react-table-plugins/blob/master/docs/useExportData.md
* Currently Single cell selection happens with ctrl + click, can we make it with click itself.
* How can a user unselect the given cells

File: examples/cell-range-selection/src/App.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  // returns two random cell ids, this is just for the demo.
+  const getRandomCellIds = React.useCallback(() => {
+    let cloneCellIds = Object.keys(cellsById)
+    let randomCellId = () =>
+      cloneCellIds[(cloneCellIds.length * Math.random()) << 0]
+    return [randomCellId(), randomCellId()]
+  }, [cellsById])
+
+  // getCellsBetweenId returns all cell Ids between two cell Id, and then setState for selectedCellIds
+  const setCells = React.useCallback(() => {
```
Comment (gargroh) [inline examples/cell-range-selection/src/App.js:76]: [minor] Renaming `setCells` -> `selectRandomCells`

File: examples/cell-range-selection/src/App.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    const cellsBetween = getCellsBetweenId(...getRandomCellIds())
+    setSelectedCellIds(cellsBetween)
+  }, [getCellsBetweenId, setSelectedCellIds, getRandomCellIds])
+
+  return (
+    <>
+      <button onClick={setCells}>setState selectedCellIds</button>
```
Comment (gargroh) [inline examples/cell-range-selection/src/App.js:83]: [minor] `setState selectedCellIds` -> `Select Cells Randomly`

File: examples/cell-range-selection/src/App.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            prepareRow(row)
+            return (
+              <tr {...row.getRowProps()}>
+                {row.cells.map(cell => {
+                  return (
+                    <td
+                      {...cell.getCellRangeSelectionProps()}
+                      {...cell.getCellProps()}
+                      style={
+                        cellsSelected[cell.id]
```
Comment (gargroh) [inline examples/cell-range-selection/src/App.js:105]: Can we add another property in `cell` itself something like isSelected(if it doesn't exists) and use `cell.isSelected` here

**Comment by gargroh** [comment]
@all-contributors please add @07harish for code example plugin

**Comment by 07harish** [comment]
> @07harish , mentioning some points let me know your thoughts on it -
> 
> * Can we have documentation of plugin something like https://github.com/gargroh/react-table-plugins/blob/master/docs/useExportData.md. Same needs to linked on https://github.com/gargroh/react-table-plugins#documentation and https://github.com/gargroh/react-table-plugins#examples
> * Currently Single cell selection happens with ctrl + click, can we make it with click itself, so as to make single selection more predictable/discoverable to user.
> * How can a user unselect the given cells

1) Yes, I will write documentation and link it

2) Single-cell selection using click (no ctrl) - Clicking outside the selected cells will **Unselect all cells by default** (this is to cancel selected cells or a range). if we make a single-cell select by click, We will not be able to provide this canceling. Even if we provide two features like cell selection and range selection, we will again have to think about the cancellation. 

3)  - (using code) Unselecting **range** and **cells** can be done using `getCellsBetweenId (if the range has to be unselected)` and then clone the `state.selectedCellsById`, modify and update using `setSelected`
     - (on UI) Unselecting **single cells** can be done using `ctrl+click` on a selected cell, and Unselecting **range** cannot be done yet (I think it's similar to Ag grid)

File: examples/cell-range-selection/src/App.js (hunk 5 above)
Comment (07harish) [reply examples/cell-range-selection/src/App.js:105]: Yeah, actually I tried it [Here](https://github.com/gargroh/react-table-plugins/pull/3/commits/fb36ae2fcc51227eef61e3dd0d82c15323e6d82a)

If I am correct, `row.allCells (all visible cells)` property is added in the `prepareRow`, So I get `undefined` in my `useInstance hook`. 

Adding this in `prepareRow` will work but, it runs only at the last and I get to see on UI after I end my selecting the range. 

I am not sure if I am adding it to the right hook (`useInstance`). So I am unable to add it. Let me know your thoughts and how I should be approaching this.

Also, As `useInstance` will trigger on each cell selects, adding `isSelected` runs loop 500 times (for eg. 100rows * 5 columns)

File: src/plguins/useCellRangeSelection.js [hunk 6]
```diff
@@ -211,24 +211,10 @@ function useInstance (instance) {
     [allColumns, cellsById, cellIdSplitBy, rows]
   )
 
-  // const selectedFlatCells = React.useMemo(()=> {
-  //   const selectedFlatCells = []
-  //   rows.forEach(row=> {
-  //     row.allCells && row.allCells.forEach(cell => {
-  //       cell.isSelected = !!selectedCellIds[cell.id]
```
Comment (07harish) [inline src/plguins/useCellRangeSelection.js]: Tried to add `selectedFlatCells` with `cell.isSelected`.

`selectedFlatCells ` had `cell.isSelected` in it, but while we loop `row.cells` on UI file (app.js), it was missing

File: examples/cell-range-selection/src/App.js (hunk 1 above)
Comment (07harish) [reply examples/cell-range-selection/src/App.js:122]: added

File: examples/cell-range-selection/README.md (hunk 2 above)
Comment (07harish) [reply examples/cell-range-selection/README.md]: updated the link, codesandbox will be available after merge hopefully.

File: examples/cell-range-selection/src/App.js (hunk 4 above)
Comment (07harish) [reply examples/cell-range-selection/src/App.js]: added

File: examples/cell-range-selection/src/App.js (hunk 3 above)
Comment (07harish) [reply examples/cell-range-selection/src/App.js]: renamed.

File: src/plguins/useCellRangeSelection.js [hunk 7]
```diff
@@ -211,10 +211,26 @@ function useInstance (instance) {
     [allColumns, cellsById, cellIdSplitBy, rows]
   )
 
+  // const selectedFlatCells = React.useMemo(()=> {
```
Comment (07harish) [inline src/plguins/useCellRangeSelection.js]: Tried to add `selectedFlatCells` with `cell.isSelected.`

`selectedFlatCells ` had `cell.isSelected` in it, but while we loop `row.cells` on UI file (app.js), it was missing

File: src/plguins/useCellRangeSelection.js [hunk 8]
```diff
@@ -211,10 +211,26 @@ function useInstance (instance) {
     [allColumns, cellsById, cellIdSplitBy, rows]
   )
```
Comment (07harish) [inline src/plguins/useCellRangeSelection.js:213]: Tried to add `selectedFlatCells` with `cell.isSelected.`

`selectedFlatCells ` had `cell.isSelected` in it, but while we loop `row.cells` on UI file (app.js), it was missing

**Comment by 07harish** [comment]
Sorry, I will update docs soon, and then you can review.

**Comment by 07harish** [comment]
Added Docs, please review and I can start writing tests.

File: examples/cell-range-selection/src/App.js (hunk 5 above)
Comment (gargroh) [reply examples/cell-range-selection/src/App.js:105]: >adding isSelected runs loop 500 times (for eg. 100rows * 5 columns) 
ahh..if that's the problem...then lets not go with tht approach...

**Comment by gargroh** [comment]
This is awesome @07harish, i am merging this PR, for test cases lets open another PR.

**Comment by 07harish** [comment]
> This is awesome @07harish, i am merging this PR, for test cases lets open another PR.

Woohooo!! 🎉🎉

**Pull request merged**